### PR TITLE
Adds mongo importer and edits existing IO modules

### PR DESCRIPTION
**Pull request by athityakumar**
- [x] Removes auto-init of class instance variables
- [x] Adds Mongo importer with tests and YARD docs
- [x] Edits YARD doc for existing IO modules
- [x] Changes suggested from PR #21 & PR #16 
- [x] Shifts `shared_examples` that are used by only one importer, to corresponding `_spec.rb` file

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,12 +5,18 @@ rvm:
   - '2.1'
   - '2.2'
   - '2.3.0'
-  - '2.4.0'
+  - '2.4.1'
```
Comment (athityakumar) [inline .travis.yml]: The Mongo gem faces issue of `unexpected argument, expected Proc` - which is due to [this bug](https://bugs.ruby-lang.org/issues/13107) in MRI Ruby 2.4.0 itself.  This has been fixed in v2.4.1. Hence, RSpec tests fail for v2.4.0 and are passing on v2.4.1.

However, I don't like the idea of changing Ruby version to 2.4.1 on Travis just because of one importer. Can the mongo importer be "silenced" / set as "pending" for just Ruby 2.4.0? Or would keeping the 2.4.1 version over 2.4.0 in Travis, be a better alternative?

File: daru-io.gemspec [hunk 2]
```diff
@@ -29,6 +29,7 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'dbi'
   spec.add_development_dependency 'jsonpath'
   spec.add_development_dependency 'mechanize'
+  spec.add_development_dependency 'mongo'
```
Comment (athityakumar) [inline daru-io.gemspec]: (Changes for Optional dependency is planned to be made later, according to discussion on issue #25 )

File: spec/daru/io/importers/html_spec.rb [hunk 3]
```diff
@@ -1,4 +1,9 @@
-RSpec.describe Daru::IO::Importers::HTML do # rubocop:disable Metrics/BlockLength
+RSpec.shared_examples 'html importer' do |symbol|
```
Comment (zverok) [inline spec/daru/io/importers/html_spec.rb]: ...and what is so specific to HTML about it?..

File: spec/daru/io/importers/csv_spec.rb [hunk 4]
```diff
@@ -1,4 +1,9 @@
-RSpec.describe Daru::IO::Importers::CSV do # rubocop:disable Metrics/BlockLength
+RSpec.shared_examples 'csv importer' do
```
Comment (athityakumar) [inline spec/daru/io/importers/csv_spec.rb]: Nothing really, but let this (issue #24) and passing readable spec like 
```Ruby
    it_behaves_like 'exact dataframe',
        order: ...,
        index: ...,
        data : ...,
        nrows: ...,
        ncols: ...
```
 be left for a separate PR. Probably in PR #27 😄

File: lib/daru/io/importers/json.rb [hunk 5]
```diff
@@ -67,103 +55,50 @@ class JSON < Base
...
-        def parse_without_hash(json)
-          @data += json
+        def fetch_data
+          return @json if @columns.empty? && @named_columns.empty?
+          return at_jsonpath(@columns.first) if @columns.size == 1 && @named_columns.empty?
+          data_columns = @columns + @named_columns.values
+          data_columns.map { |col| at_jsonpath(col) }
         end
       end
     end
```
Comment (athityakumar) [inline lib/daru/io/importers/json.rb]: Definitely yes 🎉  😄

File: Gemfile [hunk 6]
```diff
@@ -1,2 +1,13 @@
 source 'https://rubygems.org'
 gemspec
+
+group :optional do
+  gem 'activerecord', '~> 4.0'
+  gem 'dbd-sqlite3'
+  gem 'dbi'
```
Comment (athityakumar) [inline Gemfile]: This way,
- `bundle install` : Installs the gemspec gems, as well as ALL the optional parser gems.
- `bundle install --without optional` : Installs only gemspec gems.

Does this work?

File: lib/daru/io/importers/sql.rb [hunk 7]
```diff
@@ -33,14 +32,20 @@ class SQL < Base
...
+          @dbh   = dbh
+          @query = query
         end
 
         def call
+          optional_gem 'dbd-sqlite3', requires: 'dbd/SQLite3'
+          optional_gem 'activerecord', '~> 4.0', requires: 'active_record'
+          optional_gem 'dbi'
+          optional_gem 'sqlite3'
+
```
Comment (athityakumar) [inline lib/daru/io/importers/sql.rb]: Is this use-case of `Daru::IO::Base#optional_gem` readable enough?

File: lib/daru/io/base.rb [hunk 8]
```diff
@@ -1,15 +1,21 @@
...
-
-        args.each do |k, v|
-          instance_variable_set("@#{k}", v)
-          define_singleton_method(k) { instance_variable_get("@#{k}") }
-        end
+      def optional_gem(dependency, version=nil, requires: nil)
+        gem dependency, version
+        require requires || dependency
+      rescue LoadError
+        statement =
```
Comment (athityakumar) [inline lib/daru/io/base.rb]: Is this good enough? That is, should the rescued `LoadError` raise a `LoadError` again with a better error message or are there any better alternatives (like a simple `raise`, or maybe even custom error class?)

File: lib/daru/io/importers/mongo.rb [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+          orders    = @client[@collection].find.map(&:keys)
+          documents = orders.map.with_index { |order, i| Hash[order.zip data[i]] }
+
+          JSON.new(
+            documents,
+            *@columns,
+            order: @order,
+            index: @index,
+            **@named_columns
+          ).call
```
Comment (athityakumar) [inline lib/daru/io/importers/mongo.rb]: Acknowledged. I was initially apprehensive of why it'd be better for `Mongo` Importer to inherit `JSON` Importer rather than just calling it. However, requirements like raising an error when both `order` & `named_columns` were enough to justify me that inheritance would be a much DRYer way to go about it. 😄 

I've made the changes. 👍

File: lib/daru/io/base.rb (hunk 8 above)
Comment (zverok) [inline lib/daru/io/base.rb]: I believe it is good enough, only thing I'd propose (maybe later) to add the name or description of exact importer or exporter requiring the gem.

File: lib/daru/io/exporters/base.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
+require 'daru/io/base'
+
+module Daru
+  module IO
+    module Exporters
+      class Base < Base
```
Comment (zverok) [inline lib/daru/io/exporters/base.rb]: `Base < Base`?.. WAT.

File: lib/daru/io/exporters/base.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        def initialize(dataframe)
+          if dataframe.is_a?(Daru::DataFrame)
+            @dataframe = dataframe
+          else
+            raise ArgumentError,
+              'Expected first argument to be a Daru::DataFrame, '\
+              "received #{dataframe.class} instead."
+          end
```
Comment (zverok) [inline lib/daru/io/exporters/base.rb]: Turn it other way:
```ruby
raise unless dataframe.is_a?(DataFrame)
@dataframe = dataframe
```

File: lib/daru/io/exporters/sql.rb [hunk 12]
```diff
@@ -26,11 +25,16 @@ class SQL < Base
...
+          super(dataframe)
+          @dbh       = dbh
+          @table     = table
         end
 
         def call
-          require 'dbi'
+          optional_gem 'dbd-sqlite3', requires: 'dbd/SQLite3'
+          optional_gem 'dbi'
+          optional_gem 'sqlite3'
```
Comment (zverok) [inline lib/daru/io/exporters/sql.rb]: Probably better to do in `initialize`, before any other initialization/option checking.

File: lib/daru/io/importers/mongo.rb [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+          @json_input    = []
+
+          super(@json_input, *columns, order: order, index: index, **named_columns)
+        end
+
+        def call
+          optional_gem 'mongo'
+
+          client      = get_client(@connection)
+          @json_input = client[@collection].find.to_json
```
Comment (zverok) [inline lib/daru/io/importers/mongo.rb]: So, what about large collections and paging?..

File: lib/daru/io/exporters/base.rb (hunk 10 above)
Comment (athityakumar) [inline lib/daru/io/exporters/base.rb]: It's `Daru::IO::Importers::Base` inheriting from `Daru::IO::Base`. 

`Importers::Base` acts as parent class to all importers, providing importer-specific helper functions.

`Base` acts as parent class to `Importer::Base` and `Exporter::Base`< providing generic functions like `optional_gem`. Should the code be changed be explicitly specify `Base < Daru::IO::Base`?
